### DISCUSSION TITLE
Should our unit tests use `unwrap`

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3031] [General]
Does using `expect` in unit test add any real value over just using `unwrap()`?

I always thought so but in attempting to write a comment to #3019 I realized that the line number is shown so unless there are multiple panics on a single line the the `expect` is perhaps not that valuable and it is definitely more noisy.

An example using  `unwrap`: (In a unit test put `let _ = LockTime::from_seconds_ceil(0xffff_f000).unwrap();`)

```bash
failures:

---- blockdata::locktime::relative::tests::satisfied_by_height stdout ----
thread 'blockdata::locktime::relative::tests::satisfied_by_height' panicked at bitcoin/src/blockdata/locktime/relative.rs:445:58:
called `Result::unwrap()` on an `Err` value: TimeOverflowError { seconds: 4294963200 }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Same failure using `expect`: : (let _ = LockTime::from_seconds_ceil(0xffff_f000).expect(from_seconds_ceil failed);`)

```bash
failures:

---- blockdata::locktime::relative::tests::satisfied_by_height stdout ----
thread 'blockdata::locktime::relative::tests::satisfied_by_height' panicked at bitcoin/src/blockdata/locktime/relative.rs:445:58:
from_seconds_ceil failed: TimeOverflowError { seconds: 4294963200 }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by tcharding** [comment]
In researching this a bit more I found this

> A lot of confusion around unwrap(), I think, comes from well meaning folks saying things like “don’t use unwrap(),” when what they actually mean is “don’t use panicking as an error handling strategy.” This is doubly confused by a different set of people who do actually literally mean [“don’t use unwrap()”, ever, in any circumstance, to the point that it shouldn’t have existed in the first place](https://www.thecodedmessage.com/posts/2022-07-14-programming-unwrap/). This is triply confused by yet another set of people that say “don’t use unwrap(),” but actually mean, “don’t use unwrap(), expect(), slice indexing or any other panicking routine even if one proves that panicking is impossible.”

ref: https://blog.burntsushi.net/unwrap/

**Comment by apoelstra** [comment]
I should preface this by saying that I really don't care, and I don't even care enough to be consistent. But I found the "never use unwrap" blog extremely unconvincing. It more-or-less argues that `expect` makes your codebase worse but this is good because otherwise you'd panic "too much". And then he fails to understand the obvious objection that array indexing etc also panic but don't make your code super ugly, and yet he's okay with this.

**Comment by Kixunil** [comment]
In some tests I can more easily understand what went wrong seeing `assert_eq` output as opposed to other panic. This is of course not entirely related but it should be taken as inspiration.

I think `unwrap` in tests is OK, but sometimes `expect` may be better.

**Reply by apoelstra** [reply]
Definitely agreed that where possible, we should use `assert_eq` or `assert_ne` vs other kinds of panics. I wish there were others for >= and <=.
